src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs

```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
+            {
+                posts[i] = PopulateLargeObject(1);
+                posts[i - 1][0].Value.PrimaryTopic.RelatedTopics = posts[i];
+            }
+            posts[length - 1][0].Value.PrimaryTopic.RelatedTopics = posts[0];
+
+            JsonSerializerOptions options = new JsonSerializerOptions()
+            {
+                MaxDepth = depthFactor * 64,
+                IgnoreNullValues = true
+            };
+
+            Assert.Throws<JsonException> (() => JsonSerializer.Serialize(posts[0], options));
+
+            using (var memoryStream = new MemoryStream())
+            {
+                await Assert.ThrowsAsync<JsonException>(async () => await JsonSerializer.SerializeAsync(memoryStream, posts[0], options));
+            }
+        }
+
+
+        [Theory]
+        [InlineData(128)]
+        [InlineData(1024)]
+        [InlineData(4096)]
+        [InlineData(8192)]
+        [InlineData(16384)]
+        [InlineData(65536)]
+        public static async void FlushThresholdTest(int bufferSize)
+        {
+            int thresholdSize = (int)(bufferSize * 0.9 - 2);
+            FlushThresholdTestClass serialaizeObject = new FlushThresholdTestClass(PopulateObjectWhichProduceJsonWithGivenSize(bufferSize));
+            List<object> list = new List<object>();
+            StringBuilder builder = new StringBuilder();
+            for (int i = 0; i < thresholdSize; i++)
+            {
+                builder.Append("a");
+            }
+            list.Add(builder.ToString());
+            serialaizeObject.StringProperty = builder.ToString();
+            list.Add(serialaizeObject);
+            JsonSerializerOptions options = new JsonSerializerOptions();
+            options.DefaultBufferSize = bufferSize;
+
+            string json = JsonSerializer.Serialize(list);
+
+            using (var memoryStream = new MemoryStream())
+            {
+                await JsonSerializer.SerializeAsync(memoryStream, list, options);
+                string jsonSerialized = Encoding.UTF8.GetString(memoryStream.ToArray());
+                Assert.Equal(json, jsonSerialized);
+
+                List<object> deserializedList = JsonSerializer.Deserialize<List<object>>(json, options);
+                Assert.Equal(builder.ToString(), ((JsonElement)deserializedList[0]).GetString());
+                JsonElement.ObjectEnumerator obj = ((JsonElement)deserializedList[1]).EnumerateObject();
+                obj.MoveNext();               
+                Assert.Equal(builder.ToString(), obj.Current.Value.GetString());
+            }
+        }
+
+        private class FlushThresholdTestClass {
+            public string StringProperty { get; set; }
+            public List<int> ListOfInts { get; set; }
+            public FlushThresholdTestClass(List<int> list)
+            {
+                ListOfInts = list;
+            }
+        }
+
+        private static List<int> PopulateObjectWhichProduceJsonWithGivenSize(int size)
```

Comment:
"size" here seems ambiguous... I expected it to be the JSON output in bytes (or "runes"); but it's the number of elements in an array.  `GenerateListOfSize(int size)` seems better, perhaps?